File: AtCoder/ACP/BootCamp4b/Hard100_2/cpp/D_Insertion.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

ll N;
string S;

void solve() {
  ll left = 0;
  ll right = 0;
  for (int i = 0; i < N; ++i) {
    if (S[i] == '(') {
      ++left;
    } else {
      if (left > 0) {
        --left;
      } else {
        ++right;
      }
    }
  }
  string res;
  for (int i = 0; i < right; ++i) {
    res += '(';
  }
  res += S;
  for (int i = 0; i < left; ++i) {
    res += ')';
  }
  printf("%s\n", res.c_str());
}

int main() {
  cin >> N >> S;
  solve();
}

File: AtCoder/ACP/BootCamp4b/Hard100_2/python/D_Insertion.py
N = int(input())
S = input()
left, right = 0, 0
for s in S:
    if s == '(':
        left += 1
    else:
        if left > 0:
            left -= 1
        else:
            right += 1
res = '(' * right + S + ')' * left
print(res)
